modules/index/src/main/java/org/apache/ignite/internal/index/IndexBuildTask.java

```diff
@@ -301,7 +324,12 @@ private CompletableFuture<TxState> resolveFinalTxStateIfNeeded(UUID transactionI
 
         ZonePartitionId commitGroupId = new ZonePartitionId(commitPartitionId.commitZoneId, commitPartitionId.commitPartitionId);
 
-        return finalTransactionStateResolver.resolveFinalTxState(transactionId, commitGroupId);
+        return finalTransactionStateResolver.resolveFinalTxState(transactionId, commitGroupId)
+                .thenApply(txState -> {
+                    taskListener.onWriteIntentResolved(taskId, txState);
+
+                    return txState;
```

Comment:
You're free to modify the API so that `onWriteIntentResolved` returns the `txState`, and then simplify this code. And if you get rid of `taskId` then it'll become
```
.thenApply(taskListener::onWriteIntentResolved)
```
which is the most desired option in my opinion.